sa/sa.go

```diff
@@ -1549,9 +1573,168 @@ func (ssa *SQLStorageAuthority) GetOrder(ctx context.Context, req *sapb.OrderReq
 	}
 	order.Names = reversedNames
 
+	// Calculate the status for the order
+	status, err := ssa.statusForOrder(ctx, order)
+	if err != nil {
+		return nil, err
+	}
+	order.Status = &status
+
 	return order, nil
 }
 
+// statusForOrder examines the status of a provided order's authorizations to
+// determine what the overall status of the order should be. In summary:
+//   * If any of the order's authorizations are invalid, the order is invalid.
+//   * If any of the order's authorizations are deactivated, the order is deactivated.
+//   * If any of the order's authorizations are pending, the order is pending.
+//   * If all of the order's authorizations are valid, and there is
+//     a certificate serial, the order is valid.
+//   * If all of the order's authorizations are valid, and we have began
+//     processing, but there is no certificate serial, the order is processing.
+//   * If all of the order's authorizations are valid, and we haven't begun
+//     processing, then the order is pending waiting a finalization request.
+// An error is returned for any other case.
+func (ssa *SQLStorageAuthority) statusForOrder(ctx context.Context, order *corepb.Order) (string, error) {
```

Comment:
It looks like `statusForOrder` only needs `ctx` for the purpose of `getAllOrderAuthorizations`. What if we moved that call into the calling function, and made `statusForOrder` operate purely on a `*corepb.Order`? That would make the function simpler to analyze and test. We could even turn it into a function instead of a method, making it clear that it does not depend on any fields of `SQLStorageAuthority`.

Reply:
>  What if we moved that call into the calling function, and made statusForOrder operate purely on a *corepb.Order?

@jsha I don't think this will work - `*corepb.Order`'s `Authorizations` field is a `[]string` of IDs. I could make `statusForOrder` accept a `*corepb.Order` and a `[]core.Authorization` but that seems less readable for the small benefit of eliminating a `context.Context` argument. What do you think?

Reply:
Aha, I missed that. Given that complication, I think it's reasonable to keep as-is.